8 Core -> There are 8 physical processors in the chip
16 Thread-> these 8 cores are divided into 2 threads each at the software level i.e., not at ,physical level

Let's consider three tasks-
1. Game(GTA V) -> 4 Cores
2. Chrome -> 2 Cores
3. VS Code -> 2 Cores


Here Chrome Is using 2 Cores ->
So, 
1. To execute chorme -> 1 Core
2. To download and run the tabs -> 1 Core

Using Muliprocessing

In [2]:
import multiprocessing as m

In [7]:
def print_cube(num):
    print(f"Cube of {num} is {num ** 3}")
def print_square(num):
    print(f"Square of {num} is {num ** 2}")

if __name__ == "__main__":
    p1 = m.Process(target = print_square, args = (10,))
    p2 = m.Process(target = print_cube, args = (10,))

    #print("Starting the first Process: ")
    p1.start()

    #print("Starting the second process: ")
    p2.start()

    #print("Waiting for process 1 to be finished")
    p1.join()

    #print("Waiting for my second process to be finished")
    p1.join()

    #print("Both process done")


Starting the first Process: 
Starting the second process: 
Waiting for process 1 to be finished
Waiting for my second process to be finished
Both process done


In [8]:
# importing the multiprocessing module
import multiprocessing
import os

def worker1():
	# printing process id
	print("ID of process running worker1: {}".format(os.getpid()))

def worker2():
	# printing process id
	print("ID of process running worker2: {}".format(os.getpid()))

if __name__ == "__main__":
	# printing main program process id
	print("ID of main process: {}".format(os.getpid()))

	# creating processes
	p1 = multiprocessing.Process(target=worker1)
	p2 = multiprocessing.Process(target=worker2)

	# starting processes
	p1.start()
	p2.start()

	# process IDs
	print("ID of process p1: {}".format(p1.pid))
	print("ID of process p2: {}".format(p2.pid))

	# wait until processes are finished
	p1.join()
	p2.join()

	# both processes finished
	print("Both processes finished execution!")

	# check if processes are alive
	print("Process p1 is alive: {}".format(p1.is_alive()))
	print("Process p2 is alive: {}".format(p2.is_alive()))


ID of main process: 18848
ID of process p1: 2620
ID of process p2: 21024
Both processes finished execution!
Process p1 is alive: False
Process p2 is alive: False


18848 is sub-divided into 2620 and 21024 as the process id

In Multiprocessiong, a newly created process will do 2 things- 
1. run independently
2. have their own memory space

In [9]:
import multiprocessing

# empty list with global scope
result = []

def square_list(mylist):
	"""
	function to square a given list
	"""
	global result
	# append squares of mylist to global list result
	for num in mylist:
		result.append(num * num)
	# print global list result
	print("Result(in process p1): {}".format(result))

if __name__ == "__main__":
	# input list
	mylist = [1,2,3,4]

	# creating new process
	p1 = multiprocessing.Process(target=square_list, args=(mylist,))
	# starting process
	p1.start()
	# wait until process is finished
	p1.join()

	# print global result list
	print("Result(in main program): {}".format(result))


Result(in main program): []


In short p1 will have blank memory space but main process will have some memory occupied

Shared memory-
1. Array
2. value

In [10]:
import multiprocessing

def square_list(mylist, result, square_sum):
	"""
	function to square a given list
	"""
	# append squares of mylist to result array
	for idx, num in enumerate(mylist):
		result[idx] = num * num

	# square_sum value
	square_sum.value = sum(result)

	# print result Array
	print("Result(in process p1): {}".format(result[:]))

	# print square_sum Value
	print("Sum of squares(in process p1): {}".format(square_sum.value))

if __name__ == "__main__":
	# input list
	mylist = [1,2,3,4]

	# creating Array of int data type with space for 4 integers
	result = multiprocessing.Array('i', 4)

	# creating Value of int data type
	square_sum = multiprocessing.Value('i')

	# creating new process
	p1 = multiprocessing.Process(target=square_list, args=(mylist, result, square_sum))

	# starting process
	p1.start()

	# wait until the process is finished
	p1.join()

	# print result array
	print("Result(in main program): {}".format(result[:]))

	# print square_sum Value
	print("Sum of squares(in main program): {}".format(square_sum.value))


Result(in main program): [0, 0, 0, 0]
Sum of squares(in main program): 0


MULTITHREADING

Chrome -> Using 2 Cores
1. To execute Chrome -> 1 Core
2. To download and run tabs -> 1 Core

To download and run the tabs -> 1 core
1. one Thread will be used for Download
1. one Thread will be used to run the tabs



In [12]:
# Python program to illustrate the concept
# of threading
# importing the threading module
import threading

def print_cube(num):
	"""
	function to print cube of given num
	"""
	print("Cube: {}".format(num * num * num))

def print_square(num):
	"""
	function to print square of given num
	"""
	print("Square: {}".format(num * num))

if __name__ == "__main__":
	# creating thread
	t1 = threading.Thread(target=print_square, args=(10,))
	t2 = threading.Thread(target=print_cube, args=(10,))

	print("starting thread 1")
	t1.start()
	print("starting thread 2")
	t2.start()

	print("wait until thread 1 is completely executed")
	t1.join()
	print("wait until thread 2 is completely executed")
	t2.join()

	# both threads completely executed
	print("Done!")


starting thread 1
Square: 100
starting thread 2
Cube: 1000
wait until thread 1 is completely executed
wait until thread 2 is completely executed
Done!


Printing the thread id and thread name

In [14]:
# Python program to illustrate the concept
# of threading
import threading
import os

def task1():
	print("Task 1 assigned to thread: {}".format(threading.current_thread().name))
	print("ID of process running task 1: {}".format(os.getpid()))

def task2():
	print("Task 2 assigned to thread: {}".format(threading.current_thread().name))
	print("ID of process running task 2: {}".format(os.getpid()))

if __name__ == "__main__":

	# print ID of current process
	print("ID of process running main program: {}".format(os.getpid()))

	# print name of main thread
	print("Main thread name: {}".format(threading.current_thread().name))

	# creating threads
	t1 = threading.Thread(target=task1, name='thread 1 executed')
	t2 = threading.Thread(target=task2, name='thread 2 executed')

	# starting threads
	t1.start()
	t2.start()

	# wait until all threads finish
	t1.join()
	t2.join()


ID of process running main program: 18848
Main thread name: MainThread
Task 1 assigned to thread: thread 1 executed
ID of process running task 1: 18848
Task 2 assigned to thread: thread 2 executed
ID of process running task 2: 18848


In Race Condition if we try to change the varibale like following - 
1. var1 = 10 and at the same time var1 is changed to 20 i.e. now, var1 = 20
2. So might be the case the var1 will show wrong data

In [18]:
# Python program to illustrate
# the concept of race condition
# in multiprocessing
import multiprocessing

# function to withdraw from account
def withdraw(balance):	
	for _ in range(10000):
		balance.value = balance.value - 1

# function to deposit to account
def deposit(balance):	
	for _ in range(10000):
		balance.value = balance.value + 1

def perform_transactions():

	# initial balance (in shared memory)
	balance = multiprocessing.Value('i', 100)

	# creating new processes
	p1 = multiprocessing.Process(target=withdraw, args=(balance,))
	p2 = multiprocessing.Process(target=deposit, args=(balance,))

	# starting processes
	p1.start()
	p2.start()

	# wait until processes are finished
	p1.join()
	p2.join()

	# print final balance
	print("Final balance = {}".format(balance.value))

if __name__ == "__main__":
	for _ in range(10):

		# perform same transaction process 10 times
		perform_transactions()


Final balance = 100
Final balance = 100
Final balance = 100
Final balance = 100
Final balance = 100
Final balance = 100
Final balance = 100
Final balance = 100
Final balance = 100
Final balance = 100


Random Result for the above code should look like this should be like this - 
1. Final balance = 1311
2. Final balance = 199
3. Final balance = 558
4. Final balance = -2265
5. Final balance = 1371
6. Final balance = 1158
7. Final balance = -577
8. Final balance = -1300
9. Final balance = -341
10. Final balance = 157

LOCKS

In [19]:
# Python program to illustrate
# the concept of locks
# in multiprocessing
import multiprocessing

# function to withdraw from account
def withdraw(balance, lock):	
	for _ in range(10000):
		lock.acquire()
		balance.value = balance.value - 1
		lock.release()

# function to deposit to account
def deposit(balance, lock):	
	for _ in range(10000):
		lock.acquire()
		balance.value = balance.value + 1
		lock.release()

def perform_transactions():

	# initial balance (in shared memory)
	balance = multiprocessing.Value('i', 100)

	# creating a lock object
	lock = multiprocessing.Lock()

	# creating new processes
	p1 = multiprocessing.Process(target=withdraw, args=(balance,lock))
	p2 = multiprocessing.Process(target=deposit, args=(balance,lock))

	# starting processes
	p1.start()
	p2.start()

	# wait until processes are finished
	p1.join()
	p2.join()

	# print final balance
	print("Final balance = {}".format(balance.value))

if __name__ == "__main__":
	for _ in range(10):

		# perform same transaction process 10 times
		perform_transactions()


Final balance = 100
Final balance = 100
Final balance = 100
Final balance = 100
Final balance = 100
Final balance = 100
Final balance = 100
Final balance = 100
Final balance = 100
Final balance = 100
